File: Exercise_1.cpp
/* 
    Time Complexity                              :  O(N)
    Space Complexity                             :  O(1)
    Did this code successfully run on Leetcode   :  Yes
    Any problem you faced while coding this      :  No
*/

#include <bits/stdc++.h> 
using namespace std; 

// https://leetcode.com/problems/kth-smallest-element-in-a-bst

class Solution {
private:
    int ans;int ks;
public:
    int kthSmallest(TreeNode* root, int k) {
        ks = k;
        ans = -1;
        inorder(root);
        return ans;
    }
    
    void inorder(TreeNode *root) {
        if(!root) return;
        
        inorder(root->left);
        ks--;
        if(ks==0) {
            ans = root->val;
            return;
        }
        if(ans >= 0) return;
        inorder(root->right);
    } 
};

File: Exercise_1.py
#Time Complexity : O(N)), N being the number of values in tree. 
#Space Complexity : O(h), h being the height of the tree. 
#Did this code successfully run on Leetcode : Yes.
#Any problem you faced while coding this : Forgot can use the constructor to save values. 
#Your code here along with comments explaining your approach in three sentences only
'''Use in order traversal and a counter in the constructor. If counter equals k, save the 
root value into result in constructor. Return result in the end. 
'''
class Solution:
    def __init__(self):
        self.result = -1
        self.count = 0
        self.k = -1
    def helper(self, root):
        if root:
            self.helper(root.left)
            self.count = self.count+1
            if self.count==self.k:
                self.result = root.val
            self.helper(root.right)   

    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        self.k = k
        self.helper(root)
        return self.result